File: src/model/IWebSocket.cpp
/*

  INScore Project

  Copyright (C) 2015  Grame

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

  Grame Research Laboratory, 9 rue du Garet, 69001 Lyon - France
  research@grame.fr

*/

#include "IWebSocket.h"
#include "WebSocketController.h"

using namespace std;

namespace inscore
{

const string IWebSocket::kIWebSocketType("websocket");
const char * IWebSocket::kGetImgMsg = "getImage";

//-------------------------------------------------------------------------------
IWebSocket::IWebSocket(const std::string &name, IObject *parent)
	: IObject (name, parent), fWebServer(0), fPort(0), fFrequency(0)
{
	fTypeString = kIWebSocketType;
	fGetMsgHandlerMap[kstatus_GetMethod]	= TGetParamMethodHandler<IWebSocket, string (IWebSocket::*)() const>::create(this, &IWebSocket::status);
}

IWebSocket::~IWebSocket()
{
	delete fWebServer;
}

//-------------------------------------------------------------------------------
VObjectView* IWebSocket::getView() const
{
	const IObject* obj = (const IObject*)getScene();
	return obj ? obj->getView() : 0;
}

//-------------------------------------------------------------------------------
void IWebSocket::setFrequency (int freq)
{
	fFrequency = freq;
	fWebServer->setFrequency(freq);
}

//-------------------------------------------------------------------------------
bool IWebSocket::running () const		{ return fWebServer->running(); }
string IWebSocket::status () const		{ return running() ? "running" : "stopped"; }
bool IWebSocket::start (int port)		{ return fWebServer->start(port); }

//-------------------------------------------------------------------------------
MsgHandler::msgStatus IWebSocket::set (const IMessage* msg)
{
	MsgHandler::msgStatus status = IObject::set(msg);
	if (status & (MsgHandler::kProcessed + MsgHandler::kProcessedNoChange)) return status;

	// Two parameters are mandatory : port and notification time.
	if (msg->size() != 3) return MsgHandler::kBadParameters;

	int port, frequency;
	if (!msg->param(1, port) || !msg->param(2, frequency))
		return MsgHandler::kBadParameters;
	if (!port || !frequency)
		return MsgHandler::kBadParameters;

	if (frequency != fFrequency)
		setFrequency(frequency);

	if (port != fPort) {
		setPort(port);
		if ( !start(port) ) {
			ITLErr << "Cannot create web socket server on port " << fPort << ITLEndl;
			return MsgHandler::kCreateFailure;
		}
	}
	return MsgHandler::kProcessed;
}

SIMessageList IWebSocket::getSetMsg () const
{
	SIMessageList outmsgs = IMessageList::create();
	SIMessage msg = IMessage::create(getOSCAddress(), kset_SetMethod);
	*msg << kIWebSocketType << fPort << fFrequency;
	outmsgs->list().push_back (msg);
	return outmsgs;
}

//--------------------------------------------------------------------------
SIMessageList IWebSocket::getMsgs(const IMessage* msg) const
{
	SIMessageList outMsgs = IMessageList::create();
	for ( int i = 0 ; i < msg->size() ; i++ )
	{
		string param = "-";
		msg->param(i, param);

		if ( param == "clients" )
		{
			SIMessage msg = IMessage::create(getOSCAddress(), param);
			*msg << fWebServer->clients();
			outMsgs->list().push_back (msg);
		}
	}
	outMsgs->list().push_back (IObject::getMsgs(msg)->list());
	return outMsgs;
}

}

File: src/javascript/src/model/IWebSocket.ts
///<reference path="IObject.ts"/>

interface WebSocketInformer {
		getView(): VObjectView;
		getFrequency(): number;
		getPort(): number;
		//getParseEnv(): TParseEnv;
}

interface WebSocketController {
		start(port: number): boolean;
		stop()      : void;
		running()   : boolean;
		clients()   : number;
}
/*
interface TParseEnv {
	getJSEngine(): TJSEngine;
	getLUAEngine(): TLua;
	error(): void;
	parseEnv(): TParseEnv;
}
*/


class IWebSocket extends IObject {
	private fWebServer  : WebSocketController;
    protected fPort     : number;
    protected fFrequency: number;

    public kIWebSocketType  : string;
    public kMethodKey       : string;
    public kIdKey           : string;
    public kStatusKey       : string;
    public kVersionKey      : string;

    constructor(name: string, parent: IObject) {
        super(name, parent);
        this.fTypeString = kWebSocketType;
        
        super.setHandlers();
    }



	setControler(ctrl: WebSocketController): void	{ this.fWebServer = ctrl; }
	
    getPort()       : number    { return this.fPort; }
	getFrequency()  : number    { return this.fFrequency; }  

    //getParseEnv()   : TParseEnv	{ return this.getAppl(); }

    running()   : boolean		{ return this.fWebServer.running(); }
    status()    : string		{ return this.running() ? "running" : "stopped"; }
    
    start(port: number) : boolean   { return this.fWebServer.start(port); }    



    set(msg: IMessage): eMsgStatus {
        let status = super.set(msg);
        if (status & (eMsgStatus.kProcessed + eMsgStatus.kProcessedNoChange)) return status;

        // Two parameters are mandatory : port and notification time.
        if (msg.size() != 3) return eMsgStatus.kBadParameters;

        let port = -1;
        let frequency = -1;

        let a = msg.paramNum(3), b = msg.paramNum(4);

        if (!a.correct  ||  !b.correct)     { return eMsgStatus.kBadParameters; }
        if (!a.value    ||  !b.value)       { return eMsgStatus.kBadParameters; }

        let restart = false;
        if (frequency != this.fFrequency) {
            this.fFrequency = frequency;
            restart = true;
        }

        if (port != this.fPort) {
            this.fPort = port;
            restart = true;
        }

        if(restart) {
            if ( !this.start(port) ) {
                //ITLErr << "Cannot create web socket server on port " << this.fPort << ITLEndl;
                return eMsgStatus.kCreateFailure;
            }
        }
        return eMsgStatus.kProcessed;
    }



/*
    getMsgs(msg: IMessage): IMessageList {
        let outMsgs: IMessageList;
        for ( let i = 0 ; i < msg.size() ; i++ ) {
            let param: string = "-";
            msg.param(i, param);

            if ( param == "clients" ) {
                let msg: IMessage = IMessage.create(getOSCAddress(), param);
                msg << fWebServer.clients();
                outMsgs.list().push_back (msg);
            }
        }
        outMsgs.list().push_back (IObject.getMsgs(msg).list());
        return outMsgs;
    }
*/

    getSet(): IMessage	{ 
    	return new IMessage(this.getOSCAddress(), [kset_SetMethod, this.fTypeString, this.fPort, this.fFrequency]); 
    }


}